### ISSUE TITLE
distributed training can result in FileExistsError

### ISSUE THREAD

**Issue by jspreston** [allenai/allennlp#3841]
**Describe the bug**
When running multi-GPU training using the default trainer, I now get an error like:

FileExistsError: [Errno 17] File exists: '/path/to/run_dir/log/validation'

In my setup I'm running on multiple GPUs on the same machine, so the serialization directory is on a common filesystem.  It looks as though each process's tensorboardX writer is attempting to create the 'validation' directory, and there is a race condition within the tensorboardX `directory_check` method where the directory can be created by another process between the check for existence and creation:

```
    ...
    except KeyError:
        if not os.path.exists(path):
            os.makedirs(path)
```
Running the trainer again with `-f` to force writing to the now-existing directories works if I run a second time, but this is really painful for batch job submission.

Would a PR to try creating the directories TensorboardWriter prior to creating the SummaryWriter objects be acceptable? 

**To Reproduce**
Because this is a race condition it may be difficult to reproduce, but running any distributed training may cause the error.

**Expected behavior**
Running distributed training shouldn't produce FileExistsError.

**System (please complete the following information):**
 - OS: linux
 - Python version: 3.7.5
 - tried v1.0-prerelease and master
 - PyTorch version: (if you installed it yourself)

**Comment by DeNeutoy** [comment]
@jspreston a PR would be great 🙏

**Comment by DeNeutoy** [comment]
Closed by #3843
